tgui/packages/tgui/interfaces/Fabricator.tsx

```diff
@@ -0,0 +1,337 @@
+import { useBackend, useSharedState } from '../backend';
+import { Stack, Section, Button, Input, Icon, Tabs, Dimmer } from '../components';
+import { Window } from '../layouts';
+import { Material, MaterialAmount, MaterialFormatting, Materials, MATERIAL_KEYS } from './common/Materials';
+import { Fragment } from 'inferno';
+
+type Design = {
+  name: string;
+  desc: string;
+  cost: { [K in keyof typeof MATERIAL_KEYS]: number };
+  id: string;
+  categories?: string[];
+  maxstack: number;
+  icon: string;
+};
+
+type FabricatorData = {
+  materials: Material[];
+  fab_name: string;
+  on_hold: boolean;
+  designs: { [K: string]: Design };
+  busy: boolean;
+};
+
+type AvailableMaterials = { [K in keyof typeof MATERIAL_KEYS]?: number };
+
+const SearchBar = (props, context) => {
+  const [searchText, setSearchText] = useSharedState(
+    context,
+    'search_text',
+    ''
+  );
+
+  return (
+    <Stack align="baseline">
+      <Stack.Item>
+        <Icon name="search" />
+      </Stack.Item>
+      <Stack.Item grow>
+        <Input
+          fluid
+          placeholder="Search for..."
+          onInput={(_e: unknown, v: string) => setSearchText(v.toLowerCase())}
+          value={searchText}
+        />
+      </Stack.Item>
+    </Stack>
+  );
+};
+
+export const Fabricator = (props, context) => {
+  const { act, data } = useBackend<FabricatorData>(context);
+  const { materials, fab_name, on_hold, designs, busy } = data;
+  const sortedDesigns = Object.values(designs).sort((a, b) =>
+    a.name > b.name ? 1 : 0
+  );
+  const categoryCounts: { [K: string]: number } = {};
+  const [selectedCategory, setSelectedCategory] = useSharedState(
+    context,
+    'selected_category',
+    '__ALL'
+  );
+  const [searchText, setSearchText] = useSharedState(
+    context,
+    'search_text',
+    ''
+  );
+  const [displayMatCost, setDisplayMatCost] = useSharedState(
+    context,
+    'display_material_cost',
+    true
+  );
+
+  let recipeCount = 0;
+
+  Object.values(sortedDesigns).map((design) => {
+    Object.values(design.categories || []).map((name) => {
+      categoryCounts[name] = (categoryCounts[name] || 0) + 1;
+    });
+
+    recipeCount += 1;
+  });
+
+  const availableMaterials = Object.values(data.materials || []).reduce(
+    (state, value) => {
+      state[value.name] = state[value.name] || 0 + value.amount;
+      return state;
+    },
+    {}
+  ) as AvailableMaterials;
+  const sortedMaterials = (data.materials || [])
+    .splice(0)
+    .sort((a, b) => (a.name > b.name ? 1 : 0));
+
+  // This smells. You got a better idea?
+  delete categoryCounts['initial'];
+  delete categoryCounts['core'];
```

Comment:
What is this for?

Reply:
Since category grouping is done on the client now, I'm intentionally deleting `RND_CATEGORY_INITIAL` as a group, since it's not useful on the interface. Regardless, it does look like magic, and I'll move it to a `BLACKLISTED_CATEGORIES` set instead shortly